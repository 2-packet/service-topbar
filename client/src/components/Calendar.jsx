import React from 'react';
import dateFns from 'date-fns';
import '../../../public/calendar.css';

class Calendar extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			currentMonth: new Date(),
			currentDate: new Date(),
			currentDay: new Date(),
			selectedDay: new Date()
		};
		this.onDateClick = this.onDateClick.bind(this);
		this.nextMonth = this.nextMonth.bind(this);
		this.prevMonth = this.prevMonth.bind(this);
	}

	renderDays() {
		const day = 'ddd';
		const days = [];
		const firstDate = dateFns.startOfWeek(this.state.currentMonth);
		for (let i = 0; i < 7; i++) {
			days.push(
				<th className="weekday" key={i}>
					{dateFns.format(dateFns.addDays(firstDate, i), day)}
				</th>
			);
		}
		return <tr className="weekdays-header">{days}</tr>;
	}

	currentDay() {
		return dateFns.format(this.state.currentDate, 'D');
	}

	renderCells() {
		const monthStart = dateFns.startOfMonth(this.state.currentMonth);
		const startDate = dateFns.startOfWeek(monthStart);

		const rows = [];

		let days = [];
		let day = startDate;
		let formattedDate = '';

		while (rows.length < 6) {
			for (let i = 0; i < 7; i++) {
				formattedDate = dateFns.format(day, 'D');
				const cloneDay = day;
				let currentDay = '';

				let pastMonthStyle = '';
				const prevMonth = dateFns.format(
					dateFns.subMonths(this.state.currentMonth, 1),
					'MMMM'
				);

				if (prevMonth === dateFns.format(day, 'MMMM')) {
					pastMonthStyle = 'pastMonthStyle';
				}

				let futureMonthStyle = '';
				const futureMonth = dateFns.format(
					dateFns.addMonths(this.state.currentMonth, 1),
					'MMMM'
				);

				if (futureMonth === dateFns.format(day, 'MMMM')) {
					futureMonthStyle = 'futureMonthStyle';
				}

				let pastDatesStyle = '';
				if (dateFns.compareDesc(day, dateFns.subDays(new Date(), 1)) === 1) {
					pastDatesStyle = 'pastDatesStyle';
				}
				let hoverDates = '';
				if (pastDatesStyle) {
					pastMonthStyle = '';
				} else {
					hoverDates = 'hoverDates';
				}

				if (!pastDatesStyle && !pastMonthStyle) {
					currentDay =
						formattedDate === dateFns.format(this.state.selectedDay, 'D')
							? 'selectedDay'
							: '';
				}

				if (currentDay === 'selectedDay') {
					hoverDates = '';
				}

				const classNames = `${hoverDates} ${pastDatesStyle} ${futureMonthStyle} ${pastMonthStyle} calendar-day ${currentDay}`;

				days.push(
					<td key={i}>
						<div
							className={classNames}
							onClick={() => this.onDateClick(dateFns.parse(cloneDay))}
						>
							{formattedDate}
						</div>
					</td>
				);

				day = dateFns.addDays(day, 1);
			}
			rows.push(<tr key={day}>{days}</tr>);
			days = [];
		}
		return rows;
	}

	onDateClick(day) {
		if (dateFns.compareAsc(day, dateFns.subDays(new Date(), 1)) === 1) {
			if (
				dateFns.format(day, 'D') === dateFns.format(this.state.selectedDay, 'D')
			) {
				this.setState({ selectedDay: '' });
			} else {
				this.setState({ selectedDay: day });
			}
		}
	}

	nextMonth() {
		this.setState({
			currentMonth: dateFns.addMonths(this.state.currentMonth, 1)
		});
	}

	prevMonth() {
		if (
			dateFns.format(new Date(), 'MMMM YYYY') !==
			dateFns.format(this.state.currentMonth, 'MMMM YYYY')
		) {
			this.setState({
				currentMonth: dateFns.subMonths(this.state.currentMonth, 1)
			});
		}
	}

	render() {
		return (
			<div className="calendar-wrapper">
				<div className="picker-box">
					<div className="calendar-header">
						<div>
							<div className="chevron-left" onClick={this.prevMonth} />
						</div>
						<div className="col-center">
							<div className="picker-month">
								{dateFns.format(this.state.currentMonth, 'MMMM YYYY')}
							</div>
						</div>
						<div>
							<div className="chevron-right" onClick={this.nextMonth} />
						</div>
					</div>
					<table className="calendar-table">
						<thead>{this.renderDays()}</thead>
						<tbody>{this.renderCells()}</tbody>
					</table>
				</div>
			</div>
		);
	}
}

export default Calendar;
