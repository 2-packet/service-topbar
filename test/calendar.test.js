import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import Calendar from '../client/src/components/Calendar.jsx';

configure({ adapter: new Adapter() });

it('renders correctly when there are no items', () => {
	const tree = renderer.create(<Calendar />).toJSON();
	expect(tree).toMatchSnapshot();
});

describe('Rendering of Calendar', () => {
	it('should render without crashing', () => {
		const wrapper = shallow(<Calendar />);
		const dates = wrapper.find('.calendar-day');
		expect(wrapper.find('.calendar-table').length).toBe(1);
		expect(dates.length).toBe(42);
	});
});
