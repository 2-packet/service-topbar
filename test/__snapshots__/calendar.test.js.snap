// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<div
  className="calendar-wrapper"
>
  <div
    className="picker-box"
  >
    <div
      className="calendar-header"
    >
      <div>
        <div
          className="chevron-left"
          onClick={[Function]}
        />
      </div>
      <div
        className="col-center"
      >
        <div
          className="picker-month"
        >
          May 2019
        </div>
      </div>
      <div>
        <div
          className="chevron-right"
          onClick={[Function]}
        />
      </div>
    </div>
    <table
      className="calendar-table"
    >
      <thead>
        <tr
          className="weekdays-header"
        >
          <th
            className="weekday"
          >
            Sun
          </th>
          <th
            className="weekday"
          >
            Mon
          </th>
          <th
            className="weekday"
          >
            Tue
          </th>
          <th
            className="weekday"
          >
            Wed
          </th>
          <th
            className="weekday"
          >
            Thu
          </th>
          <th
            className="weekday"
          >
            Fri
          </th>
          <th
            className="weekday"
          >
            Sat
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              28
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              29
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              30
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              1
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              2
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              3
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              4
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              5
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              6
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              7
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              8
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              9
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              10
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              11
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              12
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              13
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              14
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              15
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              16
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              17
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              18
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              19
            </div>
          </td>
          <td>
            <div
              className=" pastDatesStyle   calendar-day "
              onClick={[Function]}
            >
              20
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day selectedDay"
              onClick={[Function]}
            >
              21
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              22
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              23
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              24
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              25
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              26
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              27
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              28
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              29
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              30
            </div>
          </td>
          <td>
            <div
              className="hoverDates    calendar-day "
              onClick={[Function]}
            >
              31
            </div>
          </td>
          <td>
            <div
              className="hoverDates  futureMonthStyle  calendar-day "
              onClick={[Function]}
            >
              1
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              className="hoverDates  futureMonthStyle  calendar-day "
              onClick={[Function]}
            >
              2
            </div>
          </td>
          <td>
            <div
              className="hoverDates  futureMonthStyle  calendar-day "
              onClick={[Function]}
            >
              3
            </div>
          </td>
          <td>
            <div
              className="hoverDates  futureMonthStyle  calendar-day "
              onClick={[Function]}
            >
              4
            </div>
          </td>
          <td>
            <div
              className="hoverDates  futureMonthStyle  calendar-day "
              onClick={[Function]}
            >
              5
            </div>
          </td>
          <td>
            <div
              className="hoverDates  futureMonthStyle  calendar-day "
              onClick={[Function]}
            >
              6
            </div>
          </td>
          <td>
            <div
              className="hoverDates  futureMonthStyle  calendar-day "
              onClick={[Function]}
            >
              7
            </div>
          </td>
          <td>
            <div
              className="hoverDates  futureMonthStyle  calendar-day "
              onClick={[Function]}
            >
              8
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
